pkg/auth/csp/tanzu.go

```diff
@@ -240,6 +242,21 @@ func (h *cspLoginHandler) runLocalListener(listener net.Listener) func() {
 }
 
 func (h *cspLoginHandler) callbackHandler(w http.ResponseWriter, r *http.Request) {
+	// Lock the mutex to ensure only one request can access/update shared state at a time
+	// Note: This should be remote corner case, but we have seen cases where Chrome browser redirects
+	// and make 2 back to back requests with the same URL before the local server is exited.
+	// In such scenario, the second request would be blocked till the prior request is finished
+	// and then gets unblocked to check if the prior request already acquired token, if so return
+	// the success message else let the request go through for token exchange and return
+	// the response which would fail anyway(context would be canceled for second request)
+	h.callbackHandlerMutex.Lock()
+	defer h.callbackHandlerMutex.Unlock()
+
+	// Check if token is already set by prior request
+	if h.token != nil {
+		printSuccessMessage(w, h.orgName)
+		return
+	}
 	// token exchange should be complete once this callback handler completes execution.
 	defer h.tokenExchangeComplete()
```

Comment:
In the case of two requests, could the value of `h.tokenExchangeComplete()` here be the one created by the second request? Since the locking happens after this value is set.
If so, is it a problem that we don't call the one from the first request?

Reply:
Not sure I understand your question completely.
The first request acquires the lock, would get the token and call the `h.tokenExchangeComplete()` and then release the lock, so now the second request would get unblocked and then return immediately if the token is set, else it would try to get the token and should fail because the first request already cancelled the context by calling `h.tokenExchangeComplete()`.

Reply:
After discussing, @prkalle explained that the only function that is called more than once is `callbackHandler()`.  My comment here was because I thought code run earlier could be called twice.  Since this is not the case and that a second call to `callbackHandler()` immediately blocks waiting for the lock, then my concern is invalid.